File: scripts/prepare-images.py
#!/usr/bin/env python3
from pathlib import Path
from subprocess import run, CalledProcessError

IMAGE_EXTENSIONS = {'.svg', '.png', '.jpg', '.jpeg', '.webp'}
FRONTEND_DIR = Path(__file__).resolve().parent.parent
IMAGES_DIR = FRONTEND_DIR / 'images'
OUTPUT_FILE = FRONTEND_DIR / 'src' / 'sectionImages.js'


def is_image_file(name: str) -> bool:
    return Path(name).suffix.lower() in IMAGE_EXTENSIONS


def convert_to_webp(src: Path, dest: Path) -> None:
    try:
        run(['cwebp', '-q', '80', str(src), '-o', str(dest)], check=True, capture_output=True)
    except CalledProcessError as exc:
        raise RuntimeError(f'cwebp failed for {src}: {exc.stderr.decode().strip()}') from exc


def main() -> None:
    section_images = {}

    for section_dir in sorted([p for p in IMAGES_DIR.iterdir() if p.is_dir()], key=lambda p: p.name):
        image_files = [p.name for p in sorted(section_dir.iterdir()) if p.is_file() and p.name.lower() != 'index.json' and is_image_file(p.name)]
        section_list = []

        for index, original_name in enumerate(image_files):
            original_path = section_dir / original_name
            ext = original_path.suffix.lower()
            if ext in {'.jpg', '.jpeg'}:
                target_name = f'p{index}.webp'
                target_path = section_dir / target_name
                convert_to_webp(original_path, target_path)
                original_path.unlink()
            else:
                target_name = f'p{index}{ext}'
                target_path = section_dir / target_name
                if original_path.name != target_name:
                    original_path.rename(target_path)

            section_list.append(target_name)

        section_images[section_dir.name] = section_list

    output_text = f"const sectionImages = {section_images!r};\nexport default sectionImages;\n"
    # Convert Python quote style to valid JS JSON-like output
    output_text = output_text.replace("'", '"')
    OUTPUT_FILE.write_text(output_text, encoding='utf-8')
    print(f'Generated {OUTPUT_FILE}')


if __name__ == '__main__':
    main()

File: scripts/prepare-images.js
#!/usr/bin/env node
const fs = require('fs/promises');
const path = require('path');
const { spawnSync } = require('child_process');

const frontendDir = path.resolve(__dirname, '..');
const imagesDir = path.join(frontendDir, 'images');
const outputFile = path.join(frontendDir, 'src', 'sectionImages.js');

const IMAGE_EXTENSIONS = ['.svg', '.png', '.jpg', '.jpeg', '.webp'];

function isImageFile(filename) {
  const ext = path.extname(filename).toLowerCase();
  return IMAGE_EXTENSIONS.includes(ext);
}

function getWebpName(index) {
  return `p${index}.webp`;
}

function getTargetName(index, ext) {
  if (ext === '.jpg' || ext === '.jpeg') {
    return getWebpName(index);
  }
  return `p${index}${ext}`;
}

function convertToWebp(sourcePath, destPath) {
  const result = spawnSync('cwebp', ['-q', '80', sourcePath, '-o', destPath], {
    encoding: 'utf8'
  });

  if (result.error) {
    throw result.error;
  }

  if (result.status !== 0) {
    throw new Error(`cwebp failed for ${sourcePath}: ${result.stderr || result.stdout}`);
  }
}

async function prepareImages() {
  const entries = await fs.readdir(imagesDir, { withFileTypes: true });
  const sectionDirs = entries
    .filter(entry => entry.isDirectory())
    .map(entry => entry.name)
    .sort((a, b) => a.localeCompare(b, undefined, { numeric: true, sensitivity: 'base' }));

  const sectionImages = {};

  for (const section of sectionDirs) {
    const sectionPath = path.join(imagesDir, section);
    const files = await fs.readdir(sectionPath, { withFileTypes: true });
    const imageFiles = files
      .filter(file => file.isFile())
      .map(file => file.name)
      .filter(name => name.toLowerCase() !== 'index.json')
      .filter(name => isImageFile(name))
      .sort((a, b) => a.localeCompare(b, undefined, { numeric: true, sensitivity: 'base' }));

    const sectionList = [];

    for (let index = 0; index < imageFiles.length; index += 1) {
      const originalName = imageFiles[index];
      const originalPath = path.join(sectionPath, originalName);
      const ext = path.extname(originalName).toLowerCase();
      const targetName = getTargetName(index, ext);
      const targetPath = path.join(sectionPath, targetName);

      if (ext === '.jpg' || ext === '.jpeg') {
        convertToWebp(originalPath, targetPath);
        await fs.rm(originalPath);
      } else if (originalName !== targetName) {
        await fs.rename(originalPath, targetPath);
      }

      sectionList.push(targetName);
    }

    sectionImages[section] = sectionList;
  }

  const fileContent = `const sectionImages = ${JSON.stringify(sectionImages, null, 2)};\nexport default sectionImages;\n`;
  await fs.writeFile(outputFile, fileContent, 'utf8');
  console.log(`Generated ${outputFile}`);
}

prepareImages().catch(error => {
  console.error(error);
  process.exit(1);
});
